### ISSUE TITLE
Use Google WebRTC Adapter as a driver

### ISSUE THREAD

**Issue by soulhunter1987** [onsip/SIP.js#337]
You could use https://github.com/webrtc/adapter as a shim library and use only latest spec's syntax and automatically get support for MS Edge and future Safari versions.

**Comment by derMani** [comment]
That would indeed be great. We now have to deal with different browser implementations, especially when we use specific getUserMedia constraints. 

There are already some legacy forms of getUserMedia today.


Eventually, the adapter lib won't be necessary anymore.

**Comment by soulhunter1987** [comment]
If you require webrtc-adapter you could simplify SIP.MediaHandler and environment_browser a lot. Also I think that SIP.Hacks module should be reviewed - many hacks do not make sense today.

Currently I include webrtc-adapter before sip.js and it works succesfully in Firefox/Chromium-based browsers. But MS Edge still does not work. I figured out some basic Edge requirements:
1. Needs a TURN server because does not fully support STUN - may be fixed through UA configuration
2. Needs pc.addIceCandidate(null) after last candidate to start ICE - so I internally modified SIP.MediaHandler.setDescription() method to do it after applying remote description. But currently I get DOMException then in Edge. We are trying to solve it here https://github.com/webrtc/adapter/issues/364 You may join the discussion if you are interested in Edge support.

**Comment by woopstar** [comment]
+1 on this

**Comment by wpp** [comment]
Since Apple announced WebRTC compatibility I've been playing around with the latest [Technology Previews](https://developer.apple.com/safari/technology-preview/) and SIP.js.

I had to make quite a few changes to the library in order to get a call going

- createOffer/createAnswer SIP.Utils.promisify conflicts
- setLocalDescriptiong SIP.Utils.promisify conflicts
- addStream vs addTrack
- turn server urls vs url compatibility (https://github.com/webrtc/adapter/pull/584)

I'm wondering wether this project want's to keep shimming these calls or go the direction of the proposed adapter integration is on the road-map.

Either way I'd be happy to help out and thanks for the library.

**Comment by james-criscuolo** [comment]
Hi @wpp,
  A few people within our company also gave the technology preview a go. While I didn't do it myself, turning on the "legacy webrtc api" or something like that in the settings made SIP.js work out of the box. The question becomes whether that flag (likely hidden) will be on or off by default upon release, but either way we should definitely aim to be compatible with the newer API.

In terms of the adapter integration, I'll defer to @egreenmachine. He is currently in the middle of a [large, API-breaking pull request](https://github.com/onsip/SIP.js/pull/426) that should at least minimize the amount of work necessary to make the above work (and potentially already get rid of that shim).

**Comment by egreenmachine** [comment]
I just tested #426 in the Safari developer preview. 

The majority of issues you listed are due to `promisify`. This makes sense as it is a home grown naive implementation of `promisify`. I have removed it from the WebRTC sections of my branch and use the native promise interface. I would like to continue removing `promisify` and also `deferred`. 

I am still using `addSteam` which is [deprecated](https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/addStream). So I will be switching to `addTrack` before I merge and release.

I have already updated the way we handle turn server urls.

I have not checked IE yet, but a shim might be necessary there. 

I am not convinced that using the adapter is something we should be doing by default. We should be able to keep the library up to date with the latest spec of WebRTC and compatible with browsers. I think that the simplification of the code in my PR will help with that in the long term. 

That being said, if would like to create a `SessionDescriptionHandler` that uses the adapter and submit it as a PR, we would definitely take it. Our current plan is to release #426 as 0.8.0, once we are confident that it is stable. As mentioned, it is a big refactor that is API breaking. I am hoping to start documentation in the next few days which should help people migrate and start testing.

Thanks!

**Comment by wpp** [comment]
First of all thanks for the quick responses and the "legacy webrtc api" tip.
I'll keep an eye out for #426! If you let me know when it's in a testable state I'd happy to give it a spin and report back any issues encountered.
As far as the adapter integration with a `SessionDescriptionHandler` goes: I'm not too keen on adding adapter tbh, especially since spec and vendors are getting closer to 1.0.
I simply wanted to know what direction this project is taking, which you cleared up.

**Comment by egreenmachine** [comment]
So... after further testing it appears that [Chrome does not yet support `addTrack`](https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/addTrack#Browser_compatibility). So I will have to shim this in my WebRTC handler. I did test it in Safari and it seemed happy. There are some lingering references to `promisify` that I need to go and clean up.

**Comment by soulhunter1987** [comment]
You could leave anything like this behind, if using adapter library as proposed, because it already shims addTrack for Chrome. I think you should just use the latest WebRTC spec without ant shims and propose users to use WebRTC Adapter to achieve best interpolerability and compatibility. When those shims wont be needed anymore - developers of Adapter will remove them themselves and you would not have to worry about it.

**Comment by jlaine** [comment]
I think this issue could be closed once #473 is merged. Once this is in, SIP.js would will systematically prefer up-to-date WebRTC APIs over the legacy ones, thereby supporting Firefox/Chrome/Safari/any standars-conformant browser without additional hacks.

Users wanting to support anything exotic such as Edge should include webrtc-adapter themselves and it should "just work".
